File: Javascript/1456. Maximum Number of Vowels in a Substring of Given Length.js
/*
Given a string s and an integer k, return the maximum number of vowel letters in any substring of s with length k.
Vowel letters in English are 'a', 'e', 'i', 'o', and 'u'.
*/
/**
 * 1456. Maximum Number of Vowels in a Substring of Given Length
 * @param {string} s
 * @param {number} k
 * @return {number}
 */
var maxVowels = function(s, k) {
    const n = s.length;
    if (n < k) { return 0; }
    const vowels = new Set(['a','e','i','o','u']);
    let maxVC = 0;
    let currVC = 0;
    
    for (let i = 0; i < k; i++) {
        if (vowels.has(s[i])) {
            currVC++;
        }
    }
    maxVC = currVC;
    
    for (let i = k; i < n; i++) {
        if (vowels.has(s[i - k])) {
            currVC--;
        }
        if (vowels.has(s[i])) {
            currVC++
        }
        maxVC = Math.max(maxVC, currVC);
    }
    
    return maxVC;
};

//console.log(maxVowels("abciiidef", 3)); // Expect 3
//console.log(maxVowels("aeiou", 2)); // Expect 2
console.log(maxVowels("leetcode", 3)); // Expect 2

File: Rust/1456_Maximum_Number_of_Vowels_in_a_Substring_of_Given_Length.rs
/*
Given a string s and an integer k, return the maximum number of vowel letters in any substring of s with length k.
Vowel letters in English are 'a', 'e', 'i', 'o', and 'u'.
*/
/* 1456. Maximum Number of Vowels in a Substring of Given Length */
use std::collections::HashSet;
use std::cmp::max;
struct Solution;
impl Solution {
    pub fn max_vowels(s: String, k: i32) -> i32 {
        let n = s.len();
        let s_chars: Vec<char> = s.chars().collect();
        let letters = vec!['a', 'e', 'i', 'o', 'u'];
        let vowels: HashSet<_> = letters.iter().cloned().collect();
        let mut max_vc = 0;
        let mut curr_vc = 0;
        
        for i in 0..k as usize {
            if vowels.contains(&s_chars[i]) {
                curr_vc += 1;
            }
        }
        max_vc = curr_vc;
        
        for i in k as usize..n {
            if vowels.contains(&s_chars[i - k as usize]) {
                curr_vc -= 1;
            }
            if vowels.contains(&s_chars[i]) {
                curr_vc += 1;
            }
            max_vc = max(max_vc, curr_vc);
        }
        max_vc
    }
}

fn main() {
    //let s = String::from("abciiidef");
    //let k = 3; // Expect 3
    
    //let s = String::from("aeiou");
    //let k = 2; // Expect 2
    
    let s = String::from("leetcode");
    let k = 3; // Expect 2    
    
    let result = Solution::max_vowels(s, k);
    println!("{:?}", result);
}